src/libraries/System.Linq/src/System/Linq/ToCollection.cs

```diff
@@ -255,14 +256,37 @@ private static TSource[] HashSetToArray<TSource>(HashSet<TSource> set)
 
         private static List<TSource> HashSetToList<TSource>(HashSet<TSource> set)
         {
-            var result = new List<TSource>(set.Count);
+            int count = set.Count;
 
-            foreach (TSource item in set)
+            List<TSource> result;
+            if (count > 0)
             {
-                result.Add(item);
+                result = new List<TSource>(count);
+                Span<TSource> span = SetCountAndGetSpan(result, count);
+
+                int index = 0;
+                foreach (TSource item in set)
+                {
+                    span[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == span.Length, "All list elements were not initialized.");
+            }
+            else
+            {
+                result = new List<TSource>();
             }
```

Comment:
You can just always do `new List<TSource>(count)` and let List's ctor sort it out.  Then we don't need the else block.